src/main/java/ch/njol/skript/lang/function/FunctionRegistry.java

```diff
@@ -0,0 +1,682 @@
+					.toArray(Class<?>[][]::new));
+		}
+	}
+
+	/**
+	 * Returns an unmodifiable list of candidates for the provided function identifier.
+	 *
+	 * @param provided The provided function.
+	 * @param existing The existing functions with the same name.
+	 * @return An unmodifiable list of candidates for the provided function.
+	 */
+	private static @Unmodifiable @NotNull Set<FunctionIdentifier> candidates(
+		@NotNull FunctionIdentifier provided,
+		Set<FunctionIdentifier> existing
+	) {
+		Set<FunctionIdentifier> candidates = new HashSet<>();
+
+		candidates:
+		for (FunctionIdentifier candidate : existing) {
+			// by this point, all candidates have matching names
+
+			// if we have no provided arguments, we can match any function
+			// we can skip the rest of the checks
+			if (provided.args == null || provided.args.length == 0) {
+				candidates.add(candidate);
+				continue;
+			}
+
+			// if argument counts are not possible, skip
+			if (provided.args.length > candidate.args.length
+				|| provided.args.length < candidate.minArgCount) {
+				continue;
+			}
+
+			// if the types of the provided arguments do not match the candidate arguments, skip
+			for (int i = 0; i < provided.args.length; i++) {
+				if (!Converters.converterExists(provided.args[i], candidate.args[i])) {
+					continue candidates;
+				}
+			}
+
+			candidates.add(candidate);
+		}
+
+		if (candidates.size() <= 1 || provided.args == null || provided.args.length == 0) {
+			// if there is only one candidate,
+			// or we should match any function if provided.args == null || provided.args.length == 0,
+			// then return without trying to convert
+			return Collections.unmodifiableSet(candidates);
+		}
+
+		// let overloaded(Long, Long) and overloaded(String, String) be two functions.
+		// the code below allows overloaded(1, {_x}) to match Long, Long and avoid String, String,
+		// and allow overloaded({_x}, 1) to match Long, Long and avoid String, String
+		// despite not being an exact match in all arguments,
+		// since variables have an unknown type at runtime.
+		Iterator<FunctionIdentifier> iterator = candidates.iterator();
+		while (iterator.hasNext()) {
+			FunctionIdentifier candidate = iterator.next();
+			int argIndex = 0;
+
+			while (argIndex < provided.args.length) {
+				if (provided.args[argIndex] == Object.class) {
+					argIndex++;
+					continue;
+				}
+
+				if (provided.args[argIndex] != candidate.args[argIndex]) {
+					iterator.remove();
+					break;
+				}
+
+				argIndex++;
+			}
+		}
+
+		return Collections.unmodifiableSet(candidates);
+	}
+
+	/**
+	 * Removes a function's signature from the registry.
+	 *
+	 * @param signature The signature to remove.
+	 */
+	public void remove(@NotNull Signature<?> signature) {
+		Preconditions.checkNotNull(signature, "signature cannot be null");
+
+		String name = signature.getName();
+		FunctionIdentifier identifier = FunctionIdentifier.of(signature);
+
+		synchronized (namespaces) {
```

Comment:
I don't think you need to synchronize here. Even if it is the global namespace (which is really the only one that might have concurrent operations), the map itself should handle/synchronize it properly